src/dc_sqlite3.rs

```diff
@@ -1109,371 +891,298 @@ pub unsafe fn dc_sqlite3_get_config(
 }
 
 pub fn dc_sqlite3_get_config_int64(
     context: &Context,
     sql: &dc_sqlite3_t,
-    key: *const libc::c_char,
-    def: i64,
+    key: impl AsRef<str>,
+    def: Option<i64>,
 ) -> i64 {
-    let s = unsafe { dc_sqlite3_get_config(context, sql, key, 0 as *const libc::c_char) };
-    if s.is_null() {
-        return def;
-    }
-
-    let ret: i64 = as_str(s).parse().unwrap_or_default();
-    unsafe { free(s as *mut libc::c_void) };
-    ret
+    let ret = dc_sqlite3_get_config(context, sql, key, 0 as *const libc::c_char);
+    ret.map(|r| r.parse().unwrap_or_default())
+        .unwrap_or_else(|_| def.unwrap_or_default())
 }
 
-pub unsafe fn dc_sqlite3_try_execute(
+pub fn dc_sqlite3_try_execute(
     context: &Context,
     sql: &dc_sqlite3_t,
-    querystr: *const libc::c_char,
+    querystr: impl AsRef<str>,
 ) -> libc::c_int {
     // same as dc_sqlite3_execute() but does not pass error to ui
-    let mut success = 0;
-    let sql_state;
-    let stmt = dc_sqlite3_prepare(context, sql, querystr);
-    if !stmt.is_null() {
-        sql_state = sqlite3_step(stmt);
-        if sql_state != 101 && sql_state != 100 {
-            dc_log_warning(
-                context,
-                0,
-                b"Try-execute for \"%s\" failed: %s\x00" as *const u8 as *const libc::c_char,
-                querystr,
-                sqlite3_errmsg(sql.cobj),
-            );
-        } else {
-            success = 1
+    if let Some(stmt) = dc_sqlite3_prepare(context, sql, querystr) {
+        match stmt.execute() {
+            Ok(_) => 1,
+            Err(err) => {
+                warn!(
+                    context,
+                    0, "Try-execute for \"{}\" failed: {}", querystr, err,
+                );
+                0
+            }
         }
+    } else {
+        0
     }
-    sqlite3_finalize(stmt);
-    success
 }
 
-pub unsafe fn dc_sqlite3_get_rowid(
+pub fn dc_sqlite3_get_rowid(
     context: &Context,
     sql: &dc_sqlite3_t,
-    table: *const libc::c_char,
-    field: *const libc::c_char,
-    value: *const libc::c_char,
+    table: impl AsRef<str>,
+    field: impl AsRef<str>,
+    value: impl AsRef<str>,
 ) -> uint32_t {
     // alternative to sqlite3_last_insert_rowid() which MUST NOT be used due to race conditions, see comment above.
     // the ORDER BY ensures, this function always returns the most recent id,
     // eg. if a Message-ID is splitted into different messages.
-    let mut id = 0 as uint32_t;
-    let q3 = sqlite3_mprintf(
-        b"SELECT id FROM %s WHERE %s=%Q ORDER BY id DESC;\x00" as *const u8 as *const libc::c_char,
-        table,
-        field,
-        value,
-    );
-    let stmt = dc_sqlite3_prepare(context, sql, q3);
-    if 100 == sqlite3_step(stmt) {
-        id = sqlite3_column_int(stmt, 0) as uint32_t
+    if let Some(ref conn) = sql.conn() {
+        match conn.query_row(
+            "SELECT id FROM ? WHERE ?=? ORDER BY id DESC",
```

Comment:
i am pretty sure that a `?` left of the equal sign does not work. this was the reason i did this via mprintf() iirc